tracing: Support OpenTracing API

rnburn opened the pull request:
*Description*:
* Adds an `OpenTracingDriver` that implements `Tracing::Driver` using the OpenTracing API. With it, an OpenTracing-compliant tracer can be added into envoy by deriving from `OpenTracingDrive`r and implementing the abstract `tracer` method, reducing the amount of code that must be maintained for each tracer implementation.
* Updates the LightStep tracer to use `OpenTracingDriver`. The latest version of LightStep provides a customizable transport, so the code that implements an Envoy-specific Reporter and interacts directly with LightStep's protobuf classes was removed in order to bind less tightly to the LightStep code.
* Adds a new tracing log category and logs warning when injection/extraction fails, or the LightStep tracer generates an error message.

A few notes:
* As a step towards standardizing on a single OpenTracing header for propagation, we added optional support for single-header propagation to the LightStep tracer and switched Envoy to use that instead of its custom propagation code. Currently, LightStep's single-header propagation uses the same header as Envoy; though the expectation is that it will change to match whatever standard format emerges in the future.
* Envoy still supports custom single-header propagation (though in a more OpenTracing-friendly manner). Individual tracers can be configured to use either Envoy's single-header propagation format, the tracer's native propagation format, or both via the tracer's driver class.
* To support more efficient span context propagation when a tracer's headers belong to Envoy's special O(1) list, I added the function `Http::HeaderMap::lookup` that takes a header key and uses the `StaticLookupTable` to find the header's entry. If the tracer's headers don't belong to the O(1) list, propagation will still work; but the tracer will have to iterate through all of the entries.
* The bazel build sets CC to point to a C++ compiler. This causes cmake to fail since it expects CC to be a C compiler. Even though C compilers aren't used, I added `CC=""` to `opentracing.genrule_cmd` and `lightstep.genrule_cmd` in order to work around the problem. (Not sure if there's a better solution.)

Fixes #1196 

*Risk Level*: High

*Testing*:
* Built and ran automated tests both locally on OSX and through the docker ci targets.
* Set up a small tracing [example](https://github.com/rnburn/envoy-lightstep-example) and confirmed that traces are reported to LightStep.
* Added additional unit tests for the new `Http::HeaderMap::lookup` function.

mattklein123 commented:
Thanks @rnburn. This is great. I will spend time with this tomorrow morning.

mattklein123 commented:
> The bazel build sets CC to point to a C++ compiler. This causes cmake to fail since it expects CC to be a C compiler. Even though C compilers aren't used, I added CC="" to opentracing.genrule_cmd and lightstep.genrule_cmd in order to work around the problem. (Not sure if there's a better solution.)

@jmillikin-stripe this is related I think to our other conversation yesterday about the warnings. If possible can you review the bazel portion of this change? @htuch also? I will review the code.

jmillikin-stripe commented:
That opentracing/lightstep build stuff looks painful. Lets see if we can convince them to put some BUILD files in their repo. Their libraries don't look complicated, so hopefully they'll be OK with it.

I'll send them PRs and link them here.

jmillikin-stripe commented:
OpenTracing PR: https://github.com/opentracing/opentracing-cpp/pull/37

rnburn commented:
Thanks. I'm fine with supporting bazel builds in those projects. I just merged in the PR for opentracing.

mattklein123 commented on source/common/tracing/opentracing_driver_impl.h:
```diff
@@ -0,0 +1,45 @@
+  SpanPtr spawnChild(const Config& config, const std::string& name, SystemTime start_time) override;
+
+private:
+  bool use_single_header_propagation_;
```
I think this and what follows can be const

mattklein123 replied:
Why can't this be an enum? Do we need 2 booleans?

rnburn replied:
I could change to an enum, but I would think the options could vary independently for a tracer (doesn't zipkin, for example, propagate using both methods?)

mattklein123 replied:
Perhaps, but would it ever run in both modes? I think it would be driven by config in which one mode is chosen. Enum feels more logical to me but I will admit I don't have all the context you do.

rnburn replied:
Changed to use an enum.

mattklein123 commented on source/common/tracing/opentracing_driver_impl.h:
```diff
@@ -0,0 +1,45 @@
+private:
+  bool use_single_header_propagation_;
+  bool use_tracer_propagation_;
+  std::unique_ptr<opentracing::Span> span_;
```
is there any way to avoid a double allocation for each span? I haven't looked at all at the OT C++ code yet.

rnburn replied:
It currently also allocating twice since the stored [lightstep::Span](https://github.com/envoyproxy/envoy/blob/master/source/common/tracing/lightstep_tracer_impl.h#L44) is just an [std::shared_ptr](https://github.com/lightstep/lightstep-tracer-cpp/blob/v0_38/src/c%2B%2B11/lightstep/span.h#L54).

I think the double allocation comes from having two levels of abstraction (both at the OT level and within envoy).

It could be eliminated by redesigning Envoy's Span class to be a lightweight stack allocated object that manages a pointer to either an opentracing span or a zipkin span and implements dynamic dispatch via a conditional instead of a vtable. But I don't know how elegant that is.

jmillikin-stripe commented:
@rnburn Thanks, that's great to know! I've sent another PR for Lightstep's C++ tracer: https://github.com/lightstep/lightstep-tracer-cpp/pull/85

The BUILD files for Lightstep are a bit strange because it vendors part of `googleapis`, and it assumes that the Lightstep proto headers are in the system include path. It'd be nice to clean that weirdness up at some point.

mattklein123 commented on source/common/tracing/opentracing_driver_impl.cc:
```diff
@@ -0,0 +1,168 @@
+  std::unique_ptr<opentracing::SpanContext> parent_span_ctx;
+  if (use_single_header_propagation && request_headers.OtSpanContext()) {
+    std::string parent_context = Base64::decode(request_headers.OtSpanContext()->value().c_str());
+    std::istringstream iss(parent_context);
```
stringstreams have fairly poor perf characteristics last I checked. Is this required?

rnburn replied:
I'll looking into adding binary propagation directly into/from string-like objects to the OT API.

rnburn replied:
Put in https://github.com/opentracing/opentracing-cpp/pull/42 to add serialization support directly to/from strings.

mattklein123 commented on source/common/tracing/opentracing_driver_impl.cc:
```diff
@@ -0,0 +1,168 @@
+    if (parent_span_ctx_maybe) {
+      parent_span_ctx = std::move(*parent_span_ctx_maybe);
+    } else {
+      ENVOY_LOG(warn, "Failed to extract span context: {}",
```
This could log at extremely high rate due to user supplied input. This should be debug level and also increment a stat somewhere.

rnburn replied:
Added stats to track inject/extract errors.

mattklein123 commented on source/common/tracing/opentracing_driver_impl.cc:
```diff
@@ -0,0 +1,168 @@
+    if (parent_span_ctx_maybe) {
+      parent_span_ctx = std::move(*parent_span_ctx_maybe);
+    } else {
+      ENVOY_LOG(warn, "Failed to extract span context: {}",
```
same comment about logging/stat

mattklein123 commented on source/common/tracing/opentracing_driver_impl.cc:
```diff
@@ -0,0 +1,168 @@
+  }
+  active_span = tracer.StartSpan(operation_name, {opentracing::ChildOf(parent_span_ctx.get()),
+                                                  opentracing::StartTimestamp(start_time)});
+  if (active_span == nullptr) {
```
When does this ever return nullptr?

mattklein123 commented on source/common/tracing/opentracing_driver_impl.cc:
```diff
@@ -0,0 +1,168 @@
+      return opentracing::make_unexpected(opentracing::key_not_found_error);
+    case Http::HeaderMap::Lookup::NotSupported:
+      return opentracing::make_unexpected(opentracing::lookup_key_not_supported_error);
+    default:
```
default should not be required, you have all cases covered, compiler will check

rnburn replied:
Without the default, I'll get this error when making the debug build:
```
source/common/tracing/opentracing_driver_impl.cc: In member function 'virtual opentracing::expected<opentracing::v1::string_view> Envoy::Tracing::{anonymous}::OpenTracingHTTPHeadersReader::LookupKey(opentracing::v1::string_view) const':
source/common/tracing/opentracing_driver_impl.cc:53:3: error: control reaches end of non-void function [-Werror=return-type]
   }
```

mattklein123 replied:
just remove the default, and put the NOT_REACHED at the end of the function.

rnburn replied:
Removed default case.

mattklein123 commented on source/common/tracing/opentracing_driver_impl.cc:
```diff
@@ -0,0 +1,168 @@
+SpanPtr OpenTracingSpan::spawnChild(const Config&, const std::string& name, SystemTime start_time) {
+  std::unique_ptr<opentracing::Span> ot_span = span_->tracer().StartSpan(
+      name, {opentracing::ChildOf(&span_->context()), opentracing::StartTimestamp(start_time)});
+  if (ot_span == nullptr) {
```
when does this return nullptr?

rnburn replied:
Most of the OpenTracing API is noexcept. LightStep's StartSpan would only return nullptr if something like `std::bad_alloc` was thrown while it was trying to allocate memory. I could change this to use RELEASE_ASSERT since nullptr would likely mean the application is on the verge of crashing.

mattklein123 replied:
Yeah preference to either use RELEASE_ASSERT for these types of things or in the case of nullptr just let Envoy crash via null-ptr dereference.

mattklein123 commented on source/common/tracing/opentracing_driver_impl.cc:
```diff
@@ -0,0 +1,168 @@
+    const opentracing::expected<void> was_successful =
+        span_->tracer().Inject(span_->context(), oss);
+    if (!was_successful) {
+      ENVOY_LOG(warn, "Failed to inject span context: {}", was_successful.error().message());
```
Same comments as below on error handling (sorry comments are coming out of order). In general, error handling should:
1) not log above debug if it can happen via user supplied input
2) stat
3) if it's something that just shouldn't happen, it should be an ASSERT or possibly RELEAS_ASSERT (please see error handling section in style guide)

rnburn replied:
Changed to log at debug level, but still need to add a stat.

mattklein123 commented on source/common/tracing/opentracing_driver_impl.cc:
```diff
@@ -0,0 +1,168 @@
+void OpenTracingSpan::injectContext(Http::HeaderMap& request_headers) {
+  if (use_single_header_propagation_) {
+    // Inject the span context using Envoy's single-header format.
+    std::ostringstream oss;
```
same comment as below about string stream

mattklein123 commented on source/common/tracing/opentracing_driver_impl.cc:
```diff
@@ -0,0 +1,168 @@
+                parent_span_ctx_maybe.error().message());
+    }
+  } else if (use_tracer_propagation) {
+    const OpenTracingHTTPHeadersReader reader{request_headers};
```
From looking through, I think LS driver uses this path, but not the other one. Do we have Envoy test coverage to 100% on the other paths? (Please see coverage report in CI). We will need coverage for both paths.

Also, what header is the LS driver actually using in the current version when it goes through this path? x-ot-span-ctx? Or some new header? It's not super clear.

rnburn replied:
I'll add test coverage for the other path.

It's using the same x-ot-span-context header.

rnburn replied:
Added test coverage for the other code path.

jmillikin-stripe commented:
Sure, but `#include <lightstep-tracer-common/collector.pb.h>` would keep the same warning-silencing behavior without quite as much :stonk:.

mattklein123 commented on source/common/tracing/opentracing_driver_impl.h:
```diff
@@ -0,0 +1,45 @@
+namespace Envoy {
+namespace Tracing {
+
+class OpenTracingSpan : public Span, Logger::Loggable<Logger::Id::tracing> {
```
Can we have some links to docs here or just a set of comments somewhere on how a driver is actually created? It's a little confusing to me re: what the various propagation modes are, what the driver implements, how the header lookup code works (in terms of what the OT code expects), etc.

rnburn replied:
Will do.

rnburn replied:
Added documentation.

rnburn commented:
@jmillikin-stripe - yeah, the path could be changed like that. I just meant having them included with `-isystem` instead of `-I`.

rnburn commented on source/common/tracing/lightstep_tracer_impl.cc:
```diff
@@ -15,102 +15,102 @@
+                                   lightstep::CollectorMethodName());
+  message->body() = Grpc::Common::serializeBody(request);
+
+  uint64_t timeout =
```
Fixed.

rnburn commented on source/common/tracing/lightstep_tracer_impl.cc:
```diff
@@ -15,102 +15,102 @@
-  span_.SetOperationName(operation);
+    Grpc::Common::chargeStat(*driver_.cluster(), lightstep::CollectorServiceFullName(),
+                             lightstep::CollectorMethodName(), true);
+    if (!Grpc::Common::deserializeBody(response->bodyAsString(), *active_response_)) {
```
Fixed.

rnburn commented on source/common/tracing/lightstep_tracer_impl.cc:
```diff
@@ -15,102 +15,102 @@
 
-void LightStepRecorder::enableTimer() {
+void LightStepDriver::TlsLightStepTracer::enableTimer() {
   uint64_t flush_interval =
```
Fixed.

rnburn commented on source/common/tracing/lightstep_tracer_impl.cc:
```diff
@@ -127,57 +127,28 @@ LightStepDriver::LightStepDriver(const Json::Object& config,
+    tls_options.use_single_key_propagation = true;
+    tls_options.logger_sink = LightStepLogger{};
+    tls_options.max_buffered_spans = std::function<size_t()>{[this] {
+      auto result = runtime_.snapshot().getInteger("tracing.lightstep.min_flush_spans", 5U);
```
Fixed.

jmillikin-stripe commented:
I just verified that the following Bazel repository entries work as expected with the new Lightstep/OpenTracing build rules:

```python
  native.git_repository(
      name = "io_opentracing_cpp",
      remote = "https://github.com/opentracing/opentracing-cpp",
      commit = "713c15d40ae63185d2bec99bf3b03823967d7108",
  )
  native.git_repository(
      name = "com_lightstep_tracer_cpp",
      remote = "https://github.com/lightstep/lightstep-tracer-cpp",
      commit = "LIGHTSTEP_COMMIT_HERE",
  )
  native.new_git_repository(
      name = "com_lightstep_tracer_common",
      remote = "https://github.com/lightstep/lightstep-tracer-common",
      commit = "3943f3f04e3547e7e68d7900ef2efde045fa8901",
      build_file = "@com_lightstep_tracer_cpp//:lightstep-tracer-common/BUILD",
  )
  native.new_git_repository(
      name = "lightstep_vendored_googleapis",
      remote = "https://github.com/google/googleapis",
      commit = "d6f78d948c53f3b400bb46996eb3084359914f9b",
      build_file = "@com_lightstep_tracer_cpp//:lightstep-tracer-common/third_party/googleapis/BUILD",
  )
  native.bind(
      name="lightstep",
      actual="@com_lightstep_tracer_cpp//:lightstep_tracer",
  )
```

For the things that lightstep-tracer-cpp is vendoring, I used the commits in the LS readme files.

rnburn commented on source/common/tracing/opentracing_driver_impl.h:
```diff
@@ -0,0 +1,45 @@
+  SpanPtr startSpan(const Config& config, Http::HeaderMap& request_headers,
+                    const std::string& operation_name, SystemTime start_time) override;
+
+  virtual const opentracing::Tracer& tracer() const = 0;
```
Fixed.

rnburn commented on source/common/http/header_map_impl.cc:
```diff
@@ -399,6 +399,23 @@ void HeaderMapImpl::iterateReverse(ConstIterateCb cb, void* context) const {
 
+HeaderMap::Lookup HeaderMapImpl::lookup(const LowerCaseString& key,
+                                        const HeaderEntry** entry) const {
+  StaticLookupEntry::EntryCb cb = ConstSingleton<StaticLookupTable>::get().find(key.get().c_str());
```
Via a macro, each of the O(1) headers adds a [lambda funtion](https://github.com/envoyproxy/envoy/blob/master/source/common/http/header_map_impl.cc#L216) that takes `HTTPHeaderMapImpl&` as an argument. The only way I see around const_cast is to have them also add entries for a `const HTTPHeaderMapImpl&` lambda.

IMO const_cast is the better option.

mattklein123 replied:
I could go either way. I think it's slightly cleaner to have the init code add a const callback and then have a fineConst(), but I'm ok with this. I would just add a comment on the tradeoff.

rnburn replied:
Added note for const_cast.

mattklein123 replied:
nit: For the next person that comes along and asks "why const_cast" can you add a brief additional comment on the tradeoffs between this and ^.

rnburn replied:
Added explanation for choosing const_cast.

htuch commented on include/envoy/http/header_map.h:
```diff
@@ -384,6 +384,18 @@ class HeaderMap {
+  enum class Lookup { Found, NotFound, NotSupported };
+
+  /**
+   * Lookup one of the O(1) headers by key.
```
It would be clearer to refer to these as something like "predefined inline header (see ALL_INLINE_HEADERS below)". Using the complexity alone elides the semantic info here.

rnburn replied:
Changed naming.

htuch commented on source/common/tracing/lightstep_tracer_impl.cc:
```diff
@@ -15,102 +16,106 @@
-void LightStepSpan::finishSpan() { span_.Finish(); }
+void LightStepDriver::LightStepTransporter::onSuccess(Http::MessagePtr&& response) {
+  try {
+    Grpc::Common::validateResponse(*response);
```
Envoy has a gRPC client layered above `Http::AsyncClient` in https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/async_client_impl.h. Can you make use of this? This would avoid having to hand parse out gRPC messages..

rnburn replied:
`Grpc::AsyncClientImpl` can't be used with an abstract base class for the response since it [constructs](https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/async_client_impl.h#L142) an instance of the class.

I'd rather not expose LightStep's protobuf message types directly since that would provide less encapsulation and potentially break code using the library if they were to change.

mattklein123 replied:
I discussed this a while ago with @htuch and although I think we could fixup the gRPC async client to work OK with raw messages, don't worry about this for now. I would potentially just add a TODO to look into cleaning this up in the future.

rnburn replied:
Added TODO note.

htuch commented on source/common/tracing/opentracing_driver_impl.cc:
```diff
@@ -0,0 +1,168 @@
+    }
+  }
+
+  opentracing::expected<void> ForeachKey(OpenTracingCb f) const override {
```
Nit: `forEachKey`.

rnburn replied:
This function from the the OT API and since it's override a virtual function the name has to match.

htuch commented on source/common/tracing/opentracing_driver_impl.cc:
```diff
@@ -0,0 +1,168 @@
+private:
+  const Http::HeaderMap& request_headers_;
+
+  static Http::HeaderMap::Iterate header_map_callback(const Http::HeaderEntry& header,
```
Nit: `headerMapCallback`.

rnburn replied:
Renamed.

jmillikin-stripe commented on bazel/external/BUILD:
```diff
@@ -4,7 +4,6 @@ cc_library(
     name = "all_external",
     srcs = [":empty.cc"],
     deps = [
-        "@com_github_lightstep_lightstep_tracer_cpp//:lightstep",
```
This target is used to pre-build the dependencies when generating the Docker image. We still want to pre-build Lightstep and all its deps since they're pretty big.

rnburn replied:
Fixed.

jmillikin-stripe commented on bazel/repositories.bzl:
```diff
@@ -312,23 +313,36 @@ def _com_github_gcovr_gcovr():
     )
 
+def _com_github_opentracing_opentracing_cpp():
+    location = REPOSITORY_LOCATIONS[
```
Please use `_repository_impl()` here so that downstream builders of Envoy can override the dependency definition.

```python
_repository_impl("io_opentracing_cpp")
native.bind(
    name = "opentracing",
    actual = "@io_opentracing_cpp//:opentracing",
)
```

You'll want to adjust the `repository_locations.bzl` name to match.

rnburn replied:
Changed to use _repository_impl and matched names to keys in REPOSITORY_LOCATIONS.

jmillikin-stripe commented on bazel/repositories.bzl:
```diff
@@ -312,23 +313,36 @@ def _com_github_gcovr_gcovr():
-        ],
-        genrule_cmd_file = "@envoy//bazel/external:lightstep.genrule_cmd",
-        build_file = "@envoy//bazel/external:lightstep.BUILD",
+    native.git_repository(
```
Same here.

jmillikin-stripe commented on bazel/repository_locations.bzl:
```diff
@@ -25,10 +25,17 @@ REPOSITORY_LOCATIONS = dict(
-        urls = ["https://github.com/lightstep/lightstep-tracer-cpp/releases/download/v0_36/lightstep-tracer-cpp-0.36.tar.gz"],
+        commit = "c9d9215b5c3652c7eb5640903697d9a683e1df76",
+        remote = "https://github.com/lightstep/lightstep-tracer-cpp",
+        googleapis = dict(
```
It's pretty confusing to have this extra level of nesting. How about giving googleapis its own location?

rnburn replied:
Separated out. Is it ok to build target in the same function as lightstep?

jmillikin-stripe replied:
I think so. The functions only exist for human readability, and it makes sense to have all the lightstep stuff in one place.

mattklein123 commented:
@rnburn can you ping me when all the comments have been acted on and I should look again? Thank you.

rnburn commented:
Will do.

rnburn commented:
@mattklein123 -- this should be ready for another round of reviewing.

I think I addressed all the comments except for the concerns on stringstream performance. 

There were some objections on https://github.com/opentracing/opentracing-cpp/pull/42 to adding more direct binary propagation support for strings. I changed the span context extraction code to use a more efficient version of std::istream that doesn't copy the data string (unlike std::istringstream), but it's still using std::ostringstream for inject since I don't know that there's an easy way to make a faster replacement (see [StackOverflow](https://stackoverflow.com/q/47115334/4447365)).

mattklein123 commented:
@rnburn FYI I'm trying to get the 1.5.0 release out the door so I might be a bit delayed in looking at this. Sorry. It's on my backlog.

mattklein123 commented on source/common/common/utility.h:
```diff
@@ -64,6 +64,26 @@ class ProdMonotonicTimeSource : public MonotonicTimeSource {
+/**
+ * Class used for creating non-copying std::istream's. See IMemoryStream below.
+ */
+class MemoryStreamBuffer : public std::streambuf {
```
nit: Can we call this `ConstMemoryStreamBuffer` ?

rnburn replied:
Renamed.

mattklein123 commented on source/common/common/utility.h:
```diff
@@ -64,6 +64,26 @@ class ProdMonotonicTimeSource : public MonotonicTimeSource {
+ *
+ * See https://stackoverflow.com/a/13059195/4447365.
+ */
+class IMemoryStream : public virtual MemoryStreamBuffer, public std::istream {
```
nit: `InputConstMemoryStream` ?

rnburn replied:
Renamed.

mattklein123 commented on source/common/tracing/lightstep_tracer_impl.cc:
```diff
@@ -15,103 +16,109 @@
+namespace {
+class LightStepLogger : Logger::Loggable<Logger::Id::tracing> {
+public:
+  void operator()(lightstep::LogLevel level, opentracing::string_view message) const {
```
I'm not familiar with `opentracing::string_view`. Can this by passed by const reference? Same Q in a bunch of other places.

rnburn replied:
It could, but it's usually faster to take small objects like `string_view` by value.

Also, passing by value will be the convention for `std::string_view` and is what's done in the C++17 standard library. (See for example [`operator<<`](http://en.cppreference.com/w/cpp/string/basic_string_view/operator_ltlt))

mattklein123 replied:
I'm not really sure how it can be faster to pass small objects by value vs. reference (would be curious if you have doc/article I can read on that), but totally fine to leave as is if that's what C++17 will do.

rnburn replied:
The members of small objects can get split and passed in registers, so access `std::string_view::data` can be simply read from the register. If it's passed by `const std::string_view&` then it needs to go through an extra dereference to access `std::string_view::data`. You can see it in the generated assembly.

```cpp
template <typename T>
void do_not_optimize_away(T const& val) {
  asm volatile("" : : "g"(val) : "memory");
}

void f1(std::string_view s) {
  char x = *s.data();
  char y = x + (char)1;
  do_not_optimize_away(y);
}

void f2(const std::string_view& s) {
  char x = *s.data();
  char y = x + (char)1;
  do_not_optimize_away(y);
}
```
compiled with `g++ -O3 -S` shows

f1:
```
	.globl __Z2f1St17basic_string_viewIcSt11char_traitsIcEE
__Z2f1St17basic_string_viewIcSt11char_traitsIcEE:
LFB758:
	movzbl	(%rsi), %eax
	addl	$1, %eax
	ret
```
and f2:
```
	.globl __Z2f2RKSt17basic_string_viewIcSt11char_traitsIcEE
__Z2f2RKSt17basic_string_viewIcSt11char_traitsIcEE:
LFB759:
	movq	8(%rdi), %rax
	movzbl	(%rax), %eax
	addl	$1, %eax
	ret
```

Note the extra dereferencing instructions in f2.

mattklein123 replied:
Got it, that makes sense. Thanks for the explanation!

mattklein123 commented on source/common/tracing/lightstep_tracer_impl.cc:
```diff
@@ -15,103 +16,109 @@
-                                     Event::Dispatcher& dispatcher)
-    : builder_(tracer), driver_(driver) {
+LightStepDriver::TlsLightStepTracer::TlsLightStepTracer(
+    std::shared_ptr<lightstep::LightStepTracer>&& tracer, LightStepDriver& driver,
```
nit: moving shared_ptr is a little bit of an anti-pattern IMO. I would let compiler optimize this and just pass normally.

rnburn replied:
Changed.

mattklein123 commented on source/common/tracing/lightstep_tracer_impl.h:
```diff
@@ -27,87 +31,81 @@ struct LightstepTracerStats {
   Upstream::ClusterInfoConstSharedPtr cluster() { return cluster_; }
   Runtime::Loader& runtime() { return runtime_; }
   LightstepTracerStats& tracerStats() { return tracer_stats_; }
+  void setPropagationMode(PropagationMode propagation_mode) {
```
Is there any reason this can't be passed via constructor?

rnburn replied:
I can add it as a constructor argument also. I only put it there for testing to get coverage for the different versions of Inject/Extract.

mattklein123 replied:
In the tests can you use constructor to construct a new driver? IMO it's cleaner to have this be part of object creation and not settable. If it's really required for testing please call it `setPropagationModeForTest`

rnburn replied:
Changed to use constructor.

mattklein123 commented on source/common/tracing/lightstep_tracer_impl.h:
```diff
@@ -27,87 +31,81 @@ struct LightstepTracerStats {
-  // Tracer::TracingDriver
-  SpanPtr startSpan(const Config& config, Http::HeaderMap& request_headers,
-                    const std::string& operation_name, SystemTime start_time) override;
+  // Tracer::OpenTracingDriver
```
super nit: overrides (with comment) go below other functions in a specific visibility scope (so it's clear what is what).

rnburn replied:
Rearranged.

mattklein123 commented on source/common/tracing/lightstep_tracer_impl.h:
```diff
@@ -105,7 +103,7 @@ class LightStepDriver : public OpenTracingDriver {
   Runtime::Loader& runtime_;
   std::unique_ptr<lightstep::LightStepTracerOptions> options_;
-  PropagationMode propagation_mode_ = PropagationMode::TracerNative;
+  PropagationMode propagation_mode_;
```
This can be const now I think.

rnburn replied:
Done.

rnburn commented:
@mattklein123 Thanks. Improving the performance is definitely something I'm going to look into. I expect there are quite a few cases in the OpenTracing API where we can use move semantics to avoid unnecessary allocations.

moderation commented:
@rnburn what should trigger a change in the commit versions for opentracing_cpp and lightstep_tracer_cpp? Should this match to releases for the project or should there be some other trigger to update?  They are currently:

```
    io_opentracing_cpp = dict(
        commit = "550c686e0e174c845a034f432a6c31a808f5f994",
        remote = "https://github.com/opentracing/opentracing-cpp",
    ),
    com_github_lightstep_lightstep_tracer_cpp = dict(
        commit = "deb5284395075028c3e5b4eab1416fe1e597bdb7",
        remote = "https://github.com/lightstep/lightstep-tracer-cpp",
),
```
 lightstep_tracer_cpp is 12 commits ahead of the latest release and opentracing_cpp is behind the latest minor version bump but ahead of the latest release.

rnburn commented:
I put in a release for the lightstep tracer for the bazel support so that envoy is now synced to 0.6.0.

I also put in https://github.com/envoyproxy/envoy/pull/2226 to update opentracing to match the commit for version 1.2.0.

abbi-gaurav commented:
Does it also provides the capability to use jaeger as a tracing solution since it is based on OpenTracing standard?

rnburn commented:
@abbi-gaurav Not yet. Jaeger has some compatibility with Zipkin so you can report spans to it using the Zipkin tracer ([example](https://github.com/envoyproxy/envoy/tree/master/examples/jaeger-tracing)), but it won't interoperate with other jaeger tracing without [changes](https://github.com/jaegertracing/jaeger-client-go#zipkin-http-b3-compatible-header-propagation) to the setup and it won't support sampling.

I'm working on a PR (https://github.com/envoyproxy/envoy/pull/2252) that will let you dynamically load in a tracer library, which would allow you to use Jaeger's C++ client directly. I have a [small example](https://github.com/rnburn/jaeger-envoy-docker/tree/master/example) that shows how it can work, but it's still under development.
